Grade-mini - a CRT emulation shader

Dogway opened the pull request:
This is an abridged port of Grade, a popular (initially grading only) shader for RetroArch and now included in crt-guest-advanced and most preset packages like the Mega Bezel Reflection Shader, CyberLab Mega Bezel, Sonkun’s crt-guest-advanced-hd presets, etc.

I started development in 2020, over 3 years ago and in the last two months I updated it to include gamut compression, a faithful model of a CRT transfer function and a more accurate white point mapping function.

A very old (and primitive) version of the white point mapping function was also committed to PCSX2 to allow the reproduction of Japanese developed games in their original cool temperature.

For this version, called grade-mini I stripped unrelated or highly specific settings to avoid confusion and make it more approachable.

In summary it emulates composite signal round trip (conversion to Y'UV and quantization), CRT transfer function, RGB CRT Beams, CRT Saturation, CRT phosphor gamut and color temperature. For the output it implements gamut compression and color space conversion.

This commit covers and expands on the following issue, opened by me under a different alias:
[https://bugs.dolphin-emu.org/issues/8941](https://bugs.dolphin-emu.org/issues/8941)

**EDIT**: Screenshots as requested

![1080 01](https://github.com/dolphin-emu/dolphin/assets/13509598/4b7533b1-c03f-4200-b18b-5da8d0b475d5)
![1080 02](https://github.com/dolphin-emu/dolphin/assets/13509598/86e986e2-5935-4171-ad5c-678e4cdeb946)
![MGS 01](https://github.com/dolphin-emu/dolphin/assets/13509598/e3d469e2-d67c-4a7c-9988-d28c5fa09789)
![MGS 02](https://github.com/dolphin-emu/dolphin/assets/13509598/a7a22fc0-0954-4028-9f24-4547082f4f3a)
![SMG 01](https://github.com/dolphin-emu/dolphin/assets/13509598/6fbfa179-d919-4a20-98ad-102ee63cc9f6)
![SMG 02](https://github.com/dolphin-emu/dolphin/assets/13509598/f4359e03-f3ea-45ab-8ae7-7169de15d2de)
![WW ORIG](https://github.com/dolphin-emu/dolphin/assets/13509598/1724f5f7-9d09-4d5d-a2c9-fe0816b05ed5)
NTSC-J phosphors (as above examples)
![WW NTSC-J](https://github.com/dolphin-emu/dolphin/assets/13509598/a440350b-a96e-463e-b2fa-9fa647476721)
CRT Projector phosphors
![WW Proj](https://github.com/dolphin-emu/dolphin/assets/13509598/24b889cf-a0bf-4011-8178-fcdc763f4fba)
![TP03 OFF PNG](https://github.com/dolphin-emu/dolphin/assets/13509598/2d9b61aa-86d7-4cc5-97b4-430aadf44d20)
NTSC-J phosphors
![TP04 NTSC-J PNG](https://github.com/dolphin-emu/dolphin/assets/13509598/9bf5b1a4-3bb5-496c-9684-3ffc53817fef)
CRT Projector phosphors
![TP05 Projector PNG](https://github.com/dolphin-emu/dolphin/assets/13509598/fbe58743-303e-4957-a1ff-b883a3071420)

iwubcode commented:
> *It works on Vulkan and OpenGL, but fails in DX11 due to a bug in the shader translation layer being unable to translate functions with mat3 types.

What was the error you saw?

Dogway commented:
It simply ignores the shader. If you mean the log, here it is:

```
24:02:186 VideoBackends\D3DCommon\Shader.cpp:123 E[MASTER]: Warning: Failed to compile S:/Emulators/Nintendo Wii/dolphin-master-5.0-16796-x64/User/Dump/bad_ps_5_0_D3D_1.txt: Error no especificado (0x80004005)
Debug info (ps_5_0):
S:\Emulators\Nintendo Wii\dolphin-master-5.0-16796-x64\Shader@0x000000000797DD80(250,1-4): error X3000: unrecognized identifier 'mat3'
24:02:186 VideoCommon\PostProcessing.cpp:743 E[MASTER]: Warning: Failed to compile post-processing shader grade-mini

24:02:843 VideoBackends\D3DCommon\Shader.cpp:130 W[Video]: ps_5_0 compilation succeeded with warnings:
S:\Emulators\Nintendo Wii\dolphin-master-5.0-16796-x64\Shader@0x000000000907F0A0(21,18-73): warning X3571: pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them
```

The second is a warning that I will have a look on. The Dolphin version is a one year old one since I'm still on Win7, if you want to test on Windows10 that would be appreciated.

JMC47 commented:
I don't see any problems with adding new shaders, but the fact it doesn't work in D3D makes it hard to justify merging this.  I think that should be a top priority - making sure shaders we ship with Dolphin work with dolphin.

Dentomologist commented:
Given that you're on 5.0-16796, the pow() warnings might be the ones fixed in [5.0-16831](https://dolphin-emu.org/download/dev/fac66897af7336b5e2dc2e3b879effb98b30fe47/).

MayImilae commented:
You should post some screenshots!

Dogway commented:
> You should post some screenshots!

Updated OP with some screenshots and also updated the shader with some clarification notes. Also reduced saturation a bit by default since typically saturation is lost through the signal in analogous systems.

iwubcode commented:
Yes, I can only assume you are on a _very_ old Dolphin version from dev standpoints.  I am able to run the shader using the DirectX backend just fine.  I do see some errors reported in the console, for instance, I was able to see that `m_in` wasn't initialized.

I end up seeing something odd where blacks are all white on the DirectX backend (toggling the `Composite Signal Type` fixes it)... otherwise the shader seems to operate alright.

Whether we want this, I'll leave up to other contributors to decide.  I have no interest in CRT emulation personally.

Dogway commented:
_Very_ or not, I don't know. The build is 11 months old. Which DX backend?

Fixed the 'm_in' initialization issue, as for the blacks turning white (NaN issue probably) I'm not sure where to look, it's happening on the first code block, that's for sure.

> Whether we want this, I'll leave up to other contributors to decide. I have no interest in CRT emulation personally, so I can't say if this is accurate or desired but it works fine.

Yeah, some people like their squares pixels too (think native resolution here), but as said in OP 'Grade' is a staple in all [RetroArch](https://forums.libretro.com/c/retroarch-additions/retroarch-shaders) shader presets for a few years now as well as PCSX2 (a basic white point adjustment).
Maybe we would need a poll for Dolphin's shader collection because I personally find no use for toxic, chrismas or acidtrip to name a few.

AdmiralCurtiss commented:
@iwubcode So is this ready for merge? I was kinda ignoring this because of the mentioned failures on D3D but apparently that works fine?

AdmiralCurtiss commented:
(well the commits need squashing but I mean, code-wise)

Dogway commented:
Updated with some cosmetics.
Somebody has to confirm if the NaN issue is solved for D3D as I added a clamp() where I think it might had been failing.

JosJuice commented:
@dolphin-emu-bot rebuild

iwubcode commented:
> I was kinda ignoring this because of the mentioned failures on D3D but apparently that works fine?

From a high level, does this work standpoint, yes, it works.  The statement in the opening is incorrect as they are using a build from almost a _year ago_ to test.

But it still has the blacks are white issue that I mentioned earlier.

Dogway commented:
Please check now on D3D.

AdmiralCurtiss commented:
@iwubcode Can you check this again?

AdmiralCurtiss commented:
![white-d3d](https://github.com/dolphin-emu/dolphin/assets/4522237/744155a1-08d6-4915-bcd2-77061973b037)

Whites are still broken in D3D.

AdmiralCurtiss commented on Data/Sys/Shaders/grade-mini.glsl:
```diff
@@ -0,0 +1,594 @@
+DefaultValue = 2
+
+[OptionRangeInteger]
+GUIName = Diplay Color Space (0:709 1:sRGB 2:P3-D65 3:Custom (Edit L508))
```
There's a typo here, `Diplay`.

Dogway replied:
@AdmiralCurtiss Thanks for feedback, can you check with 19435? I want to rule out the HDR thingy implemented after it is messing with the whites. And if possible without texture mods.
Otherwise I can only add tighter clamps until next visit to my friend's house.

Dogway commented:
Now should be working for DX11, thank you.

AdmiralCurtiss commented:
Also produces an error when loading in OpenGL:

![grafik](https://github.com/dolphin-emu/dolphin/assets/4522237/f52c7fed-fd7e-434f-97f3-3495d5c06401)

[bad_p_OGL_1.txt](https://github.com/dolphin-emu/dolphin/files/12227224/bad_p_OGL_1.txt)

Dogway commented:
> Still broken, unfortunately. Pure blacks now render as pink in DX11.
> 

What is that, doesn't DX11 really understand 0/255 is 0? do I have to spoonfeed it all?

Try the next, and if still fails I'll remove the whole clamp() which is what exactly was working at my friend's house.

AdmiralCurtiss commented:
Nope, still wrong. I threw this into RenderDoc and you seem to be producing a `nan` (via `atan2(0, 0)`), which is not surprising that it behaves differently between backends (IIRC DirectX optimizes around assuming nans never happen).

![grafik](https://github.com/dolphin-emu/dolphin/assets/4522237/1a74623d-d652-4a3e-acbb-8cdd9e340c30)

Dogway commented:
In the version I tested which is more recent but not latest , DX11 had no problems with atan () it was all about  the Quantize function. Then I added the clamp for yuv  so the conversion makes sense , nothing else . 

So all I need there is a ternary you say ?

AdmiralCurtiss commented:
Yes, `float    hue  = col.z == 0.0 && col.y == 0.0 ? 0.0 : atan(col.z,  col.y);` fixes it. `atan2(0, 0)` is undefined so the fact that OpenGL/Vulkan worked fine was kind of a happenstance.

Dogway commented:
> Yes, `float hue = col.z == 0.0 && col.y == 0.0 ? 0.0 : atan(col.z, col.y);` fixes it. `atan2(0, 0)` is undefined so the fact that OpenGL/Vulkan worked fine was kind of a happenstance.

Thank you, updated, I think only col.z check is necessary, tested this on OpenGL and everuthing seems to be working as expected so I hope DX11 does too.

Dogway commented:
Was expecting to make it at least to the progress report so at least people have a sane alternative to the mess of https://github.com/dolphin-emu/dolphin/pull/11850

Is there any issue preventing this from being merged? This is taking longer than I expected and I need to move on.

iwubcode commented:
You being stuck on a very old version of Dolphin and therefore not being able to do proper testing for D3D has been the main blocker.  If D3D is working, I see no problem merging it.

Also, not sure how your post processing shader would have any impact on what the user is complaining about there.  That has to do with the swapchain format impacting the driver's ability to engage GSYNC as far as I can tell.

JMC47 commented:
@dolphin-emu-bot rebuild

Dogway commented:
> You being stuck on a very old version of Dolphin and therefore not being able to do proper testing (in particular for D3D) has been the main blocker. If D3D is working, I see no problem merging it.
> 
> Also, not sure how your post processing shader would have any impact on what the user is complaining about there. That has to do with the swapchain format impacting the driver's ability to engage GSYNC as far as I can tell.

The last "fixing" commit was from 7 days ago, anyway if you have followed the PR I've been testing on a friend's house with Windows10 and D3D gave no issues with a build just before the PR I linked. If that PR caused any issues as noted by AdmiralCurtis it would be dealt with the changes he suggested, so it might be related, so far it seems 10-bit framebuffer can't be disabled leading to (so far) unknown bugs.

Dogway commented:
Retiring the PR due to lack of interest.

I start classes on Tuesday and have no more time to deal with apathy and disdain attitudes. First time I experience this on a PR, 3 months and ongoing to commit a mere non-disruptive and high quality shader tried-and-tested for 3 years.

For those that are still interested, follow the next links:

My Github repo for [Dolphin shaders](https://github.com/Dogway/emulation-random/tree/master/Dolphin/Shaders), includes grade-mini and PAR_Zoom_Offset.
The original Grade [thread](https://forums.libretro.com/t/dogways-grading-shader-slang/27148) in the libretro forums.

Thanks for everything.
